### PR TITLE
fix: Allow querying of 9th, 19th, 29th, etc collections

### LINKED ISSUE #2810: Many-many bridge appears to not be working
The following test fails on the last query - `"playerRewards"` returns an empty slice instead of the related object expected:
```
func TestFooFull(t *testing.T) {
	test := testUtils.TestCase{
		Actions: []any{
			testUtils.SchemaUpdate{
				Schema: `
					############################
# SETTINGS
############################

type Settings {
  id: String # @id
  value: String
}

############################
# SESSION
############################

type Session {
  id: String # @id
  sid: String # @unique
  data: String
  expiresAt: DateTime
}

############################
# CONTACT
############################

type Contact {
  id: String # @id @default(uuid())
  email: String # @unique @db.Citext
  status: String # [ENUM] SubscriptionStatus # @default(NON_SUBSCRIBED)
  preferences: [String]
  segments: [String]
  player: Player @relation(name: "playerContact") # @relation("PlayerContact")
  createdAt: DateTime # @default(now())
  updatedAt: DateTime # @updatedAt
  deletedAt: DateTime
}

############################
# PLAYER
############################

type Player {
  id: String # @id @default(uuid())
  authId: String # @unique
  characterName: String # @unique
  email: String # @unique @db.Citext
  refCode: String # @unique
  country: String # @default("")
  banned: Boolean # @default(false)
  confirmed: Boolean # @default(false)
  acceptedTos: Boolean # @default(false)
  score: Int # @default(0)
  level: Int # @default(1)
  contact: Contact @primary @relation(name: "playerContact")
  playerActions: [PlayerAction] @relation(name: "playerPlayerActions")
  playerRewards: [PlayerReward] @relation(name: "playerPlayerRewards")
  connectedAccounts: [ConnectedAccount] @relation(name: "playerConnectedAccounts")
  playerNotification: [PlayerNotification] @relation(name: "playerNotifications")
  referrals: [Referral] @relation(name: "fromPlayerReferral")
  fromReferral: Referral @relation(name: "toPlayerReferral")
  createdAt: DateTime # @default(now())
  updatedAt: DateTime # @updatedAt
  deletedAt: DateTime
}

type Referral {
  id: String
  fromPlayer: Player @primary @relation(name: "fromPlayerReferral")
  toPlayer: Player @primary @relation(name: "toPlayerReferral")
  createdAt: DateTime # @default(now())
}

type ConnectedAccount {
  id: String # @id @default(uuid())
  provider: String
  providerKey: String # @unique
  displayName: String
  platform: String # [ENUM] Platform
  providerToken: String
  providerRefreshToken: String
  player: Player @primary @relation(name: "playerConnectedAccounts") # @relation(fields: [playerId], references: [id], onDelete: Cascade)
  profile: ConnectedAccountProfile @primary @relation(name: "connectedAccountProfiles") # ConnectedAccountProfile? @relation(fields: [profileId], references: [id])
  createdAt: DateTime # @default(now())
  updatedAt: DateTime # @updatedAt

  # @@unique([provider, providerKey], name: "provider_providerKey")
  # @@index([provider, providerKey], map: "provider_providerKey_index")
}

type ConnectedAccountProfile {
  id: String # @id @default(uuid())
  provider: String
  providerKey: String
  profile: JSON
  platform: String
  connectedAccounts: [ConnectedAccount] @relation(name: "connectedAccountProfiles")
  createdAt: DateTime # @default(now())
  updatedAt: DateTime # @updatedAt
}

type PlayerAction {
  id: String
  metadata: JSON
  deleted: Boolean
  action: Action @relation(name: "actionPlayerActions")
  player: Player @relation(name: "playerPlayerActions")
  playerRewards: [PlayerActionToPlayerReward] @relation(name: "playerActionPlayerRewards")
  # playerReward: [PlayerReward] # MANY_TO_MANY
  createdAt: DateTime

  # @@unique([actionId, createdAt, playerId, metadata], name: "action_unqiue")
  # @@index([actionId], map: "actionId_index")
  # @@index([playerId], map: "playerId_index")
}

type PlayerReward {
  id: String
  points: Int
  deleted: Boolean
  player: Player @relation(name: "playerPlayerRewards")
  reward: Reward @relation(name: "rewardPlayerRewards")
  rewardTier: RewardTier @primary @relation(name: "playerRewardRewardTiers")
  events: [PlayerEvent] @relation(name: "playerRewardPlayerEvents") # This appears unused
  playerActions: [PlayerActionToPlayerReward] @relation(name: "playerRewardPlayerActions")
  # playerActions: [PlayerAction] # MANY_TO_MANY
  createdAt: DateTime # @default(now())

  # @@index([playerId])
  # @@index([rewardId])
  # @@index([rewardTierId])
}

type PlayerActionToPlayerReward {
  playerReward: PlayerReward @primary @relation(name: "playerRewardPlayerActions")
  playerAction: PlayerAction @primary @relation(name: "playerActionPlayerRewards")
}

type PlayerEvent {
  id: String # @id @default(uuid())
  createdAt: DateTime # @default(now())
  updatedAt: DateTime # @updatedAt
  type: String # [ENUM] PlayerEventType
  metadata: JSON
  playerId: String
  playerReward: PlayerReward @primary @relation(name: "playerRewardPlayerEvents") # @relation(fields: [playerRewardId], references: [id])
}

type PlayerNotification {
  id: String # @id @default(uuid())
  notification: Notification @primary @relation(name: "notificationPlayerNotifications") # @relation(fields: [notificationId], references: [id], onDelete: Cascade)
  player: Player @primary @relation(name: "playerNotifications") # @relation(fields: [playerId], references: [id], onDelete: Cascade)
  sentAt: DateTime
  acknowledgeAt: DateTime
  meta: JSON

  # @@unique([playerId, notificationId], name: "playerId_notificationId")
  # @@index([playerId])
  # @@index([notificationId])
}

type PlayerMailEvent {
  id: String # @id @default(uuid())
  type: String
  playerId: String
  contactId: String
  data: JSON
  createdAt: DateTime # @default(now())
}

type PlayerQuizzes {
  playerId: String # @id
  data: JSON
  createdAt: DateTime # @default(now())
  updatedAt: DateTime # @updatedAt
}

type PlayerGuesses {
  id: String # @id @default(uuid())
  playerId: String
  questionId: String
  choiceId: String
  platform: String # [ENUM] Platform
  isCorrect: Boolean # @default(false)
  attempts: Int # @default(0)
  completed: Boolean # @default(false)
  createdAt: DateTime # @default(now())
}

############################
# NOTIFICATIONS
############################

type Notification {
  id: String # @id @default(uuid())
  type: String
  title: String
  description: String
  meta: JSON
  expiresAt: DateTime
  playerNotifications: [PlayerNotification] @relation(name: "notificationPlayerNotifications")

  # @@unique([type, meta], name: "type_meta")
}

############################
# ACTIONS + REWARDS
############################

type Action {
  id: String # @id @default(uuid())
  key: String # @unique @db.VarChar(30)
  name: String # @db.VarChar(100)
  description: String # @db.VarChar(250)
  enabled: Boolean # @default(true)
  levelRequired: Int # @default(0)
  actionRequirements: [ActionRequirement] @relation(name: "actionActionRequirements")
  playerActions: [PlayerAction] @relation(name: "actionPlayerActions")
  createdAt: DateTime # @default(now())

  # @@index([key])
}

type ActionRequirement {
  id: String # @id @default(uuid())
  count: Int
  isPreReq: Boolean # @default(false)
  action: Action @relation(name: "actionActionRequirements")
  reward: Reward @relation(name: "rewardActionRequirements")
  createdAt: DateTime # @default(now())

  # @@unique([actionId, rewardId], name: "actionId_rewardId")
  # @@index([actionId])
  # @@index([rewardId])
}

type Reward {
  id: String # @id @default(uuid())
  name: String # @unique
  description: String
  value: Int
  limit: Int
  tiers: [RewardTier] @relation(name: "rewardRewardTiers")
  priority: Int
  enabled: Boolean # @default(true)
  actionRequirements: [ActionRequirement] @relation(name: "rewardActionRequirements")
  playerRewards: [PlayerReward] @relation(name: "rewardPlayerRewards")
  rewardGroup: [String] # [ENUM] RewardGroup
  createdAt: DateTime # @default(now())
}

type RewardTier {
  id: String # @id @default(uuid())
  level: Int
  name: String
  description: String
  value: Int
  Reward: Reward @primary @relation(name: "rewardRewardTiers") # @relation(fields: [rewardId], references: [name])
  PlayerReward: [PlayerReward] @relation(name: "playerRewardRewardTiers")

  # @@unique([level, rewardId], name: "level_rewardId")
  # @@index([rewardId])
}

############################
# TWITTER
############################

type TwitterCursor {
  id: String # @id @default(uuid())
  cursor: String
  key: String # @unique @default("")
}

type TweetActivity {
  id: String # @default(uuid())
  userId: String
  type: String #TweetActivityType
  data: JSON
  createdAt: DateTime # @default(now())

  # @@id([id, userId, type])
}

type Tweet {
  id: String # @id
  createdAt: DateTime
  conversationId: String
  tweet: JSON
}

############################
# BOT DATA
############################

type AppRole {
  id: String # @id @default(uuid())
  serverId: String
  refId: String
  app: String # [ENUM] Application
  data: JSON
  createdAt: DateTime # @default(now())
  updatedAt: DateTime # @updatedAt

  # @@unique([app, refId, serverId], name: "app_unique")
  # @@index([refId])
}

type AppChannel {
  id: String # @id @default(uuid())
  serverId: String
  refId: String
  app: String # [ENUM] Application
  data: JSON
  createdAt: DateTime # @default(now())
  updatedAt: DateTime # @updatedAt

  # @@unique([app, refId, serverId], name: "app_unique")
  # @@index([refId])
}

type AppUser {
  id: String # @id @default(uuid())
  serverId: String
  refId: String
  app: String # [ENUM] Application
  data: JSON
  createdAt: DateTime # @default(now())
  updatedAt: DateTime # @updatedAt

  # @@unique([app, refId, serverId], name: "app_unique")
  # @@index([refId])
}

type AppInteraction {
  id: String # @id @default(uuid())
  serverId: String
  refId: String
  app: String # [ENUM] Application
  type: String # [ENUM] AppInteractionType
  data: JSON
  createdAt: DateTime # @default(now())
  updatedAt: DateTime # @updatedAt

  # @@unique([app, refId, type, serverId], name: "app_unique")
  # @@index([refId])
}

############################
# ENUMS
############################

# enum Application {
#   TELEGRAM
#   DISCORD
# }

# enum AppInteractionType {
#   MESSAGE
#   REACTION
#   INVITE
# }

# enum Platform {
#   WEB
#   BOT
# }

# enum SubscriptionStatus {
#   SUBSCRIBED
#   UNSUBSCRIBED
#   PENDING
#   ARCHIVED
#   NON_SUBSCRIBED
# }

# enum TweetActivityType {
#   TWEET
#   MENTION
#   REPLY
#   RETWEET
#   QUOTE
#   LIKE
#   HASHTAG
# }

# enum PlayerEventType {
#   TIER_UPDATE
# }

# enum RewardGroup {
#   SETUP
#   TELEGRAM
#   TWITTER
#   DISCORD
#   ACADEMY
# }
				`,
			},

			testUtils.Request{
				Request: `
						mutation create_PlayerAction {
							create_PlayerAction(input: {
							metadata: "{\"test\": \"1234\"}"
							}) {
								__typename
								_docID
								metadata
							}
						}
					`,
				Results: []map[string]any{
					{
						"__typename": "PlayerAction",
						"_docID":     "bae-beae148d-9adb-5ed6-a9bd-13fe56af3ee1",
						"metadata":   "{\"test\":\"1234\"}",
					},
				},
			},
			testUtils.Request{
				Request: `
						mutation create_PlayerReward {
							create_PlayerReward(input: {
							points: 100
						}) {
								__typename
								_docID
								points
							}
						}
					`,
				Results: []map[string]any{
					{
						"__typename": "PlayerReward",
						"_docID":     "bae-eba5c340-44ce-5a74-a02f-53d30b4fb6bf",
						"points":     100,
					},
				},
			},
			testUtils.Request{
				Request: `
						mutation create_PlayerActionToPlayerReward {
							create_PlayerActionToPlayerReward(input: {
								playerAction: "bae-beae148d-9adb-5ed6-a9bd-13fe56af3ee1",
								playerReward: "bae-eba5c340-44ce-5a74-a02f-53d30b4fb6bf"
							}) {
								__typename
								_docID
								playerAction {
									_docID
									metadata
								}
								playerReward {
									_docID
									points
								}
							}
						}
					`,
				Results: []map[string]any{
					{
						"__typename": "PlayerActionToPlayerReward",
						"_docID":     "bae-65142008-5738-57cf-bcca-04cd33227947",
						"playerAction": map[string]any{
							"_docID":   "bae-beae148d-9adb-5ed6-a9bd-13fe56af3ee1",
							"metadata": "{\"test\":\"1234\"}",
						},
						"playerReward": map[string]any{
							"_docID": "bae-eba5c340-44ce-5a74-a02f-53d30b4fb6bf",
							"points": int64(100),
						},
					},
				},
			},
			testUtils.Request{
				Request: `
							query query_PlayerAction {
								PlayerAction {
									_docID
									playerRewards {
										_docID
									}
								}
							}
						`,
				Results: []map[string]any{
					{
						"_docID": "bae-beae148d-9adb-5ed6-a9bd-13fe56af3ee1",
						"playerRewards": []map[string]any{
							{
								"_docID": "bae-65142008-5738-57cf-bcca-04cd33227947",
							},
						},
					},
				},
			},
		},
	}

	testUtils.ExecuteTestCase(t, test)
}
```
Interestingly, a trimmed down SDL produces the desired result, suggesting that something 'extra' is interfering with the properties used in the requests:
```
type PlayerAction {
	id: String
	metadata: JSON
	playerRewards: [PlayerActionToPlayerReward] @relation(name: "playerActionPlayerRewards")
}

type PlayerReward {
	id: String
	points: Int
	playerActions: [PlayerActionToPlayerReward] @relation(name: "playerRewardPlayerActions")
}

type PlayerActionToPlayerReward {
	playerReward: PlayerReward @primary @relation(name: "playerRewardPlayerActions")
	playerAction: PlayerAction @primary @relation(name: "playerActionPlayerRewards")
}
```

### PR DESCRIPTION
**Pull request by AndrewSisley**
## Relevant issue(s)

Resolves #2810

## Description

Allows querying of 9th, 19th, 29th, etc collections.

The fetcher calls `PrefixEnd` which causes the start prefix to be (e.g.) `"/9"` and the end `"/10"` causing nothing to be returned to the user.  This PR fixes that by encoding the DataStore CollectionRootIDs as variable sized BigEndians.

https://github.com/sourcenetwork/defradb/issues/2818 has been broken out of this as it is not a user-visible issue atm and resolving it is a bit more involved due to some legacy tech. debt.

### CODE REVIEW

File: tests/integration/utils2.go [hunk 1]
```diff
@@ -1301,6 +1300,7 @@ func createDocViaGQL(
 	node client.DB,
 	collections []client.Collection,
 ) ([]*client.Document, error) {
+	println(fmt.Sprint(collections))
```
Comment (islamaliev) [inline tests/integration/utils2.go:1303]: question: is it left here intentional?

File: internal/core/encoding.go [hunk 2]
```diff
@@ -288,3 +288,79 @@ func EncodeIndexDataStoreKey(key *IndexDataStoreKey) []byte {
...
+	if len(data) > 1 {
+		if data[0] == '/' {
+			data = data[1:]
+		}
+		instanceType = InstanceType(data[0])
+		data = data[1:]
+	}
+
+	var docID string
+	if len(data) > 40 {
```
Comment (islamaliev) [inline internal/core/encoding.go:317]: todo: please extract the magic number to a const

File: tests/integration/issues/2810_test.go [hunk 3]
```diff
@@ -0,0 +1,315 @@
+// Copyright 2024 Democratized Data Foundation
```
Comment (shahzadlone) [inline tests/integration/issues/2810_test.go]: Praise: Thanks a bunch for documenting these! Really nice to see the issue narrowed down with the edge cases you highlighted.

File: internal/core/encoding.go [hunk 4]
```diff
@@ -288,3 +288,79 @@ func EncodeIndexDataStoreKey(key *IndexDataStoreKey) []byte {
...
+	}
+	data = data[1:]
+
+	data, colRootID, err := encoding.DecodeUvarintAscending(data)
+	if err != nil {
+		return DataStoreKey{}, err
+	}
+
+	var instanceType InstanceType
+	if len(data) > 1 {
```
Comment (islamaliev) [inline internal/core/encoding.go:308]: suggestion: wouldn't it be better to do:
```go
if len(data) <= 1 {
  return ...
}

// check instance type ...
```
Currently if `if len(data) > 1` is false the rest of the function is irrelevant, but is being executed.

File: docs/website/guides/explain-systems.md [hunk 5]
```diff
@@ -101,8 +101,8 @@ query @explain {
                     "collectionID": "3",
                     "collectionName": "Author",
                     "spans": [{
-                        "start": "/3",
-                        "end": "/4"
+                        "start": "/\x8b",
```
Comment (shahzadlone) [inline docs/website/guides/explain-systems.md:104]: nitpick: My eyes 😮

File: tests/integration/explain/default/top_with_count_test.go [hunk 6]
```diff
@@ -124,8 +124,8 @@ func TestDefaultExplainTopLevelCountRequestWithFilter(t *testing.T) {
 							},
 							"spans": []dataMap{
 								{
-									"start": "/3",
-									"end":   "/4",
+									"start": "/\x8b",
+									"end":   "/\x8c",
```
Comment (islamaliev) [inline tests/integration/explain/default/top_with_count_test.go:128]: though: this is pain. Would be nice to have it extracted so that if it changes again, we don't need to go through it again.

File: internal/core/key_test.go [hunk 7]
```diff
@@ -34,86 +34,6 @@ func TestNewDataStoreKey_ReturnsEmptyStruct_GivenEmptyString(t *testing.T) {
 	assert.ErrorIs(t, ErrEmptyKey, err)
 }
 
-func TestNewDataStoreKey_ReturnsCollectionIdAndIndexIdAndDocIDAndFieldIdAndInstanceType_GivenFourItemsWithType(
```
Comment (shahzadlone) [inline internal/core/key_test.go:37]: question: Do we loose any coverage or paths that these tests were testing due to the removal?

File: tests/integration/query/simple/with_multiple_types_test.go [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+
+package simple
+
+import (
+	"testing"
+
+	testUtils "github.com/sourcenetwork/defradb/tests/integration"
+)
+
+func TestSimple_WithSevenDummyTypesBefore(t *testing.T) {
```
Comment (islamaliev) [inline tests/integration/query/simple/with_multiple_types_test.go:19]: nitpick: would be nice to have "expectation" part in the test name:
`TestSimple_WithSevenDummyTypesBefore_ShouldSucceed` or something like this.

File: internal/core/encoding.go [hunk 9]
```diff
@@ -288,3 +288,81 @@ func EncodeIndexDataStoreKey(key *IndexDataStoreKey) []byte {
...
+
+	return DataStoreKey{
+		CollectionRootID: uint32(colRootID),
+		InstanceType:     (instanceType),
+		DocID:            docID,
+		FieldID:          fieldID,
+	}, nil
+}
+
+func EncodeDataStoreKey(key *DataStoreKey) []byte {
```
Comment (shahzadlone) [inline internal/core/encoding.go:343]: suggestion: A line documenting the Encoding type

File: internal/core/encoding.go [hunk 10]
```diff
@@ -288,3 +288,81 @@ func EncodeIndexDataStoreKey(key *IndexDataStoreKey) []byte {
 
 	return b
 }
+
+func DecodeDataStoreKey(data []byte) (DataStoreKey, error) {
```
Comment (shahzadlone) [inline internal/core/encoding.go:292]: suggestion: A line documenting the decoding type

File: tests/integration/query/simple/with_multiple_types_test.go [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+					type Type5 {
+						f: String
+					}
+					type Type6 {
+						f: String
+					}
+
+					type User {
+						name: String
+					}
```
Comment (islamaliev) [inline tests/integration/query/simple/with_multiple_types_test.go:48]: suggestion: for these tests would be nice to have some more information (as a documentation or test description) explaining why these weird-looking tests are necessary and which edge-case they test.

**Review by shahzadlone — COMMENTED** [review]
Left some comments

File: internal/core/key_test.go [hunk 12]
```diff
@@ -34,86 +34,6 @@ func TestNewDataStoreKey_ReturnsEmptyStruct_GivenEmptyString(t *testing.T) {
...
-	instanceType := "anyType"
-	fieldId := "f1"
-	docID := "docID"
-	collectionId := "1"
-	inputString := "/db/data/" + collectionId + "/" + instanceType + "/" + docID + "/" + fieldId + "/version_number"
-
-	_, err := NewDataStoreKey(inputString)
-
-	assert.ErrorIs(t, ErrInvalidKey, err)
-}
```
Comment (islamaliev) [inline internal/core/key_test.go:115]: question: you just removed the tests. Don't you plan to replace them with new ones?

**Review by islamaliev — APPROVED** [review]
Looks good. I left some clarifying comments.

File: internal/core/key_test.go (hunk 7 above)
Comment (AndrewSisley) [reply internal/core/key_test.go:37]: I doubt it, in theory everything should be covered by integration tests anyway, and the new code has fewer branches IIRC.

File: internal/core/encoding.go (hunk 9 above)
Comment (AndrewSisley) [reply internal/core/encoding.go:343]: Will add

- [ ] doc func

File: internal/core/encoding.go (hunk 10 above)
Comment (AndrewSisley) [reply internal/core/encoding.go:292]: Will add

- [ ] doc func

File: tests/integration/utils2.go (hunk 1 above)
Comment (AndrewSisley) [reply tests/integration/utils2.go:1303]: It is not :)  And I should have been more careful so soon after the mutation type bug :)  Thanks for spotting

- [ ] remove print

File: internal/core/encoding.go (hunk 2 above)
Comment (AndrewSisley) [reply internal/core/encoding.go:317]: Will do

- [ ] Rm magic number

**Review by shahzadlone — APPROVED** [review]
Please double check when all tests are passing that the removed unit tests weren't testing any edge cases before merge, other than that all good! Thanks a bunch for sorting this out.

File: internal/core/encoding.go (hunk 4 above)
Comment (AndrewSisley) [reply internal/core/encoding.go:308]: I do prefer the simplicity of the current code, although it does have a small runtime cost. I'll have a quick look when I remove the magic number - do let me know if you have a strong preference though.

- [ ] Have quick look at early returns

File: tests/integration/explain/default/top_with_count_test.go (hunk 6 above)
Comment (AndrewSisley) [reply tests/integration/explain/default/top_with_count_test.go:128]: Explain was actually pretty easy considering how invasive they are by nature, the unit tests were much more of a pain.

This might get changed slightly before merge anyway - the http client does not like this encoding and it caught me out.

File: tests/integration/query/simple/with_multiple_types_test.go (hunk 11 above)
Comment (AndrewSisley) [reply tests/integration/query/simple/with_multiple_types_test.go:48]: Agreed, I'll add some in - thanks for pushing past my laziness here :)

- [ ] Add test docs

File: internal/core/key_test.go (hunk 12 above)
Comment (AndrewSisley) [reply internal/core/key_test.go:115]: No, I don't think they had much value before, and even less so with the new code IMO.

File: internal/core/encoding.go (hunk 4 above)
Comment (islamaliev) [reply internal/core/encoding.go:308]: I wouldn't call it a strong preference. Just following return-early best practice.

File: docs/website/guides/explain-systems.md (hunk 5 above)
Comment (AndrewSisley) [reply docs/website/guides/explain-systems.md]: Has been reverted - encoding is not json friendly, so explain must be pretty-printed anyway

File: tests/integration/explain/default/top_with_count_test.go (hunk 6 above)
Comment (AndrewSisley) [reply tests/integration/explain/default/top_with_count_test.go]: This has been reverted, the encoding was not json friendly and needed to be prettied up in order to work with the http client.

File: internal/core/encoding.go (hunk 4 above)
Comment (AndrewSisley) [reply internal/core/encoding.go:309]: I had a play, whilst I like it hightlights the data flow a little more nicely, it ends up duplicating logic (or mutating, or a new func) and even without my gut prefers the feel of the current. Leaving as-is.
